afkak/client.py

```diff
@@ -652,42 +660,61 @@ def _clean_close_dlist(result, close_dlist):
-            brokerClient = self.clients.pop(broker)
+        for brokerClient in clients:
             log.debug("Calling close on: %r", brokerClient)
-            dList.append(
-                brokerClient.close().addErrback(
-                    _log_close_failure, brokerClient))
+            d = brokerClient.close().addErrback(_log_close_failure, brokerClient)
+            dList.append(d)
         self.close_dlist = DeferredList(dList)
         self.close_dlist.addBoth(_clean_close_dlist, self.close_dlist)
 
-    def _update_brokers(self, new_brokers, remove=False):
-        """ Update our self.clients based on brokers in received metadata
+    def _update_brokers(self, brokers, remove=False):
+        """
+        Update `self._brokers` and `self.clients`
+
+        Update our self.clients based on brokers in received metadata
         Take the received dict of brokers and reconcile it with our current
         list of brokers (self.clients). If there is a new one, bring up a new
         connection to it, and if remove is True, and any in our current list
         aren't in the metadata returned, disconnect from it.
-        """
-        log.debug("%r: _update_brokers: %r remove: %r",
-                  self, new_brokers, remove)
 
-        # Work with the brokers as sets
-        new_brokers = set(new_brokers)
-        current_brokers = set(self.clients.keys())
+        :param brokers: Iterable of `BrokerMetadata`. A client will be created
+            for every broker given if it doesn't yet exist.
+        :param bool remove:
+            Is this metadata for *all* brokers? If so, clients for brokers
+            which are no longer found in the metadata will be closed.
+        """
+        log.debug("%r: _update_brokers(%r, remove=%r)",
+                  self, brokers, remove)
+        brokers_by_id = {bm.node_id: bm for bm in brokers}
+        added_brokers = set(brokers_by_id) - set(self._brokers)
+        removed_brokers = set(self._brokers) - set(brokers_by_id)
 
-        # set of added
-        added_brokers = new_brokers - current_brokers
-        # removed
-        removed_brokers = current_brokers - new_brokers
+        to_close = []
 
-        # Create any new brokers based on the new metadata
-        for broker in added_brokers:
-            self._get_brokerclient(*broker)
+        self._brokers.update(brokers_by_id)
+        for node_id in removed_brokers:
+            self._brokers.pop(node_id)
+            to_close.append(self._clients.pop(node_id))
 
-        # Disconnect and remove from self.clients any removed brokerclients
+        # Remove any clients for brokers which no longer exist.
         if remove and removed_brokers:
-            self._close_brokerclients(removed_brokers)
+            for node_id in removed_brokers:
+                to_close.append(self.clients.pop(node_id))
+
+        # Identify any clients with out of date metadata and remove them.
```

Comment:
Hmmm, I found this comment confusing at first. What you mean is a broker where the current cluster metadata has a `broker id -> (host, port)` mapping that doesn't match our KafkaBrokerClient's mapping, right?  Not sure if there's a better way to write that, but I was confused by the idea that a KBC "has" metadata...

Reply:
Yeah all of this logic is super mushy as I haven't fully thought it through yet. That'll come when I write tests for it.